FWCore/Framework/interface/UnscheduledCallProducer.h

```diff
@@ -49,7 +49,7 @@ namespace edm {
       }
     }
 
-    void setEventSetup(EventSetupImpl const& iSetup) { aux_.setEventSetup(&iSetup); }
+    void setEventSetup(EventSetupImpl const* iSetup) { aux_.setEventSetup(iSetup); }
```

Comment:
In the current version of this pull request, the pointer will be null for the ProcessBlock case.

Here is a solution. There are 3 parts to this solution.

1. I percolate the TransitionInfo through the prefetchAsync_ functions of the ProductResolvers. We discussed this already. I already mostly implemented this and then backed it out because it was slightly complicated. But I still have the code and was going back and forth in my mind about whether to use it.

2. I rewrite UnscheduledProductResolver::resolveProduct_ so that it will never call doWork. Possibly I am missing something, but I think that code path never gets executed in the current version of the Framework. A prefetch will always occur and we can throw if it did not. Do you agree? Possibly I am forgetting something?

3. Remove the EventSetupImpl entirely from UnscheduledAuxiliary. Then delete the setEventSetup functions entirely.

This also has the benefit that the need for the const_cast's vanishes entirely.

Should I proceed with this solution?

Reply:
> 2\. I rewrite UnscheduledProductResolver::resolveProduct_ so that it will never call doWork. Possibly I am missing something, but I think that code path never gets executed in the current version of the Framework. A prefetch will always occur and we can throw if it did not. Do you agree? Possibly I am forgetting something?

Your suggestion sounds plausible and reasonable to me, but how about trying it out in a separate PR (I believe it can be done separately)? Then we could test it in IBs separately from the rest (assuming it passes PR tests).  @Dr15Jones, what do you think?

Reply:
I can do this in a separate PR. It seems a reasonable suggestion to separate it. I think that means we leave the 2 const_cast's as they are in this PR and also let the setEventSetup function use a pointer argument temporarily.

Reply:
I think a separate PR makes sense.

Reply:
I was actually thinking the separate PR already in parallel to this PR. We could see if removing the `doWork()` call works, and continue in this PR accordingly.